### DISCUSSION TITLE
Accounting for `end: false` in the matcher score

### DISCUSSION THREAD

**Discussion by skirtles-code** [vuejs/router#2153] [Ideas]
While working on #2148, I ran into some problems with `end: false`.

I was trying to ensure that my changes didn't impact the matching of routes, even in weird edge cases. In most cases that proved relatively straightforward, apart from for `end: false`.

My underlying assumption is that the route ranking mechanism is supposed to ensure that 'more specific' routes come before 'less specific' routes. Generally speaking, that does seem to be how it's implemented.

My PR handles static strings in paths specially. As static strings are 'more specific' than dynamic strings, that usually works out fine, as they'd come earlier in the ranking anyway.

But `end: false` is a notable outlier. It behaves a lot like a `.*` wildcard, but it doesn't really get taken into account when calculating the score. It does impact the inclusion of the 'strict bonus' for `strict: true`, but apart from that it's ignored.

With enough work, it probably is possible to get #2148 to retain the current sort order for matchers, but it'd likely involve a lot of special handling just for `end: false`. Instead, I think it would be better to change the existing behaviour, so that `end: false` is taken into account in the score.

If this idea is acceptable in principle, I'm happy to have a go at implementing it in `pathParserRanker.ts`. This would be separate from #2148, though it should ultimately allow that PR to be completed.

**Comment by posva** [comment]
I actually wonder if anybody uses `end: false`. I initially kept it because Vue Router 3 had it in path-to-regexp but I don't find it useful by default. I think that it could be its own matcher (once it's allowed to pass a matcher to the router creation)

**Reply by posva** [reply]
I'm thinking of deprecating `end: false` since it's not possible in the new custom resolvers and it just doesn't seem useful in the context of vue router
